string.c

```diff
@@ -3303,12 +3303,27 @@ rb_str_buf_cat_ascii(VALUE str, const char *ptr)
     }
 }
 
+static inline bool
+str_enc_asciicompat_from_index(int index)
+{
+    switch (index) {
+      case ENCINDEX_ASCII:
+      case ENCINDEX_UTF_8:
+      case ENCINDEX_US_ASCII:
+        return true;
+      default:
+        return rb_enc_asciicompat(rb_enc_from_index(index));
+    }
+}
+
 VALUE
 rb_str_buf_append(VALUE str, VALUE str2)
 {
-    int str2_cr;
-
-    str2_cr = ENC_CODERANGE(str2);
+    int str2_cr = rb_enc_str_coderange(str2);
+    if (str2_cr == ENC_CODERANGE_7BIT && str_enc_asciicompat_from_index(ENCODING_GET_INLINED(str))) {
+        str_buf_cat(str, RSTRING_PTR(str2), RSTRING_LEN(str2));
+        return str;
```

Comment:
I think you have a potential issue with the resulting code range here. If the `str`'s code range is `ENC_CODERANGE_BROKEN`, the concatenation could change the string to `ENC_CODERANGE_VALID`, in which case you'd need to update (or clear) the `str`'s code range.

Reply:
Looking this over more, what you have is okay for UTF-8. UTF-8 won't use a byte < 0x80 as part of a multi-byte codepoint. But, I don't know if that's true of all ASCII-compatible encodings.

Reply:
I think it's a good argument to reduce this optimization to just the 3 "built-in" (can't find a good term for them) encodings.

Reply:
Ok, done. I called this `str_enc_fastpath`.

Right now it needs 3 int comparisons, but if we were to ensure that encindex `3` is free, we'd be able to reduce this check to a single bitwise `AND`.

But right now encindex 3 is `RUBY_ENCINDEX_UTF_16BE`. So leaving this for a potential followup, as anyway it's not even showing on profiles. It may be interesting however if this check becomes more common.

Reply:
To rephrase what @nirvdrum said (it helped me understand it): all ASCII-compatible encodings interpret a character-leading byte < 0x80 as an ASCII character (implied by ASCII-compatible), but some variable-length ASCII-compat encodings might use a byte < 0x80 as the 2nd+ byte in a multi-byte character.